### PR TITLE
[FLINK-33792] Generate the same code for the same logic

### PR DESCRIPTION
**Pull request by zoudan**
## What is the purpose of the change
This pull request ensure that we generate the same code for the same logic, it is a precondition for sharing generated classes between different jobs.

## Brief change log
  - add a name counter in each `CodeGeneratorContext` and use it when we generate  names for variables.

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:

  - Add a new test: CodeGenUtilsTest#testNewName
  - Add tests in existing class: HashCodeGeneratorTest#testHashWithIndependentNameCounter and ProjectionCodeGeneratorTest#testHashWithIndependentNameCounter

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala [hunk 1]
```diff
@@ -80,23 +80,25 @@ object HashCodeGenerator {
       }
     """.stripMargin
 
+    System.out.println(code)
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:83]: Remove it.

**Review by libenchao — CHANGES_REQUESTED** [review]
@zoudan Thanks for the PR, it looks great. My main point is that we don't need to add a config for it, it can be done by default.

**Comment by zoudan** [comment]
@libenchao Thanks for reviewing, I have remove the configuration, please take a look when you have time.

**Comment by libenchao** [comment]
@zoudan The PR looks good to me, except that the CI is failing, could you fix that?

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@libenchao CI is passed, please take a look when you have time.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala [hunk 2]
```diff
@@ -37,15 +37,28 @@ import org.apache.flink.util.InstantiationUtil
...
+ * CodeGeneratorContext(or contexts share the same ancestors) is used throughout the entire class to
+ * avoid naming conflicts. So when we create a context for a class, we can set parentCtx to null.
+ * However, when we create a context for a code block, we must ensure that all contexts for code
+ * blocks in a class share a common ancestor by setting parentCtx.
  */
-class CodeGeneratorContext(val tableConfig: ReadableConfig, val classLoader: ClassLoader) {
+class CodeGeneratorContext(
+    val tableConfig: ReadableConfig,
+    val classLoader: ClassLoader,
+    parentCtx: CodeGeneratorContext) {
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala:58]: Can this be `final` too?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 3]
```diff
@@ -124,14 +124,27 @@ object CodeGenUtils {
 
   private val nameCounter = new AtomicLong
 
-  def newName(name: String): String = {
-    s"$name$$${nameCounter.getAndIncrement}"
+  def newName(context: CodeGeneratorContext = null, name: String): String = {
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:127]: This one has a default value for `context`, but `newNames` does not have, is there any rational behind this?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 4]
```diff
@@ -124,14 +124,27 @@ object CodeGenUtils {
...
+      s"$name$$${context.getNameCounter.getAndIncrement}"
+    }
   }
 
-  def newNames(names: String*): Seq[String] = {
+  def newNames(context: CodeGeneratorContext, names: String*): Seq[String] = {
     require(names.toSet.size == names.length, "Duplicated names")
-    val newId = nameCounter.getAndIncrement
-    names.map(name => s"$name$$$newId")
+    if (context == null || context.getNameCounter == null) {
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:139]: I'm wondering that if `newNames` can be optimized to something like `names.map(name => newName(context, name))`?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/CodeGenUtilsTest.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+import scala.collection.mutable.ArrayBuffer
+
+class CodeGenUtilsTest {
+  private val classLoader = Thread.currentThread().getContextClassLoader
+
+  @Test
+  def testNewName(): Unit = {
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/CodeGenUtilsTest.scala:31]: Could you also add a test for `parentCtx`?

**Review by libenchao — COMMENTED** [review]
@zoudan Thanks for the updating, using 'parent context' is a great idea. I'll left a few minor comments, others looks good to me.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala (hunk 3 above)
Comment (zoudan) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:127]: Parameter section with *-parameter is not allowed to have default arguments

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala (hunk 4 above)
Comment (zoudan) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:139]: The returned new names use the same index, I keep it the same as before. So, I did not call `newName`

**Review by lsyldliu — COMMENTED** [review]
@zoudan Thanks for contribution, I will take a look in the next few days.

**Comment by zoudan** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala (hunk 3 above)
Comment (libenchao) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:127]: My point is that do you really need to have a default value, IIUC, we want all places to pass `context`?

**Comment by zoudan** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala (hunk 3 above)
Comment (zoudan) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala]: Get it, removed

**Review by libenchao — APPROVED** [review]
+1 from my side

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRule.java [hunk 6]
```diff
@@ -88,7 +91,16 @@ public ZoneId getSessionZoneId() {
                 public ClassLoader getClassLoader() {
                     return classLoader;
                 }
+
+                @Override
+                @Nullable
+                public CodeGeneratorContext getCodeGeneratorContext() {
+                    return null;
```
Comment (lsyldliu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRule.java:98]: Can add some annotation to explain why here returns null?

**Review by lsyldliu — COMMENTED** [review]
@zoudan Thanks for contribution, LGTM overall, I just left some minor comments.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRule.java (hunk 6 above)
Comment (zoudan) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRule.java]: Change it to non null

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/LongHashJoinGenerator.scala [hunk 7]
```diff
@@ -87,10 +90,11 @@ object LongHashJoinGenerator {
   def genProjection(
       tableConfig: ReadableConfig,
       classLoader: ClassLoader,
-      types: Array[LogicalType]): GeneratedProjection = {
+      types: Array[LogicalType],
+      parentCtx: CodeGeneratorContext): GeneratedProjection = {
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/LongHashJoinGenerator.scala:94]: I just mean that we can get the `tableConfig` and `classLoader` from `parentCtx`, but this isn't an important point, ignore it now.

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@flinkbot run azure

**Comment by zoudan** [comment]
@lsyldliu I have updated my code, please have a took when you have time.

**Review by lsyldliu — APPROVED** [review]
LGTM

**Pull request closed without merging**